concurrency/TrackingSet.java

```diff
@@ -0,0 +1,232 @@
+        eligible.add(e);
+    }
+
+    /**
+     * Moves the element to inProgress subset if it was already present in eligible.
+     */
+    public synchronized void moveToInProgress(E e) {
+        if (eligible.contains(e)) {
+            eligible.remove(e);
+            inProgress.add(e);
+        }
+    }
+
+    /**
+     * Moves the element to eligible subset if it was already present in inProgress subset.
+     */
+    public synchronized void moveToEligible(E e) {
+        if (inProgress.contains(e)) {
+            inProgress.remove(e);
+            eligible.add(e);
+        }
+    }
+
+    /**
+     * Moves the element to stale subset if it was already present in eligible subset.
+     */
+    public synchronized void moveToStale(E e) {
+        if (eligible.contains(e)) {
+            eligible.remove(e);
+            stale.add(e);
+        }
+    }
+
+    /**
+     * @return a new Set containing all the elements from eligible subset.
+     */
+    public synchronized Set<E> getEligibleSet() {
+        return new HashSet<>(eligible);
+    }
+
+    /**
+     * Returns a new set containing elements with the given positive size from eligible subset after sorting. The order
+     * of the elements may not be retained in the returned set.
+     *
+     * @param size
+     *            positive number of elements to return, if size is greater the current element count in eligible set,
+     *            entire set is returned.
+     * @param comparator
+     *            used to compare the elements in the list to decide ordering.
+     *
+     * @return the set with the give size after sorting.
+     */
+    public synchronized Set<E> getSortedEligibleSet(int size, Comparator<E> comparator) {
+        Validate.isTrue(size >= 0, "size should be positive integer, but was:" + size);
+        if (size >= eligible.size()) {
+            return getEligibleSet();
+        }
+        return new LinkedList<E>(eligible).stream().collect(
+                    Collectors.toCollection(
+                        // heaping sorting
+                        ()->MinMaxPriorityQueue.orderedBy(comparator).maximumSize(size).create()
+                    )).stream().collect(Collectors.toSet());
+    }
+
+    @Override
+    public DiagnosticInfo getDiagnosticInfo() {
+        return new SummaryDiagnostic();
+    }
+
+    /**
+     * Diagnostic information for the tracking set.
+     */
+    public final class SummaryDiagnostic extends RootDiagnosticInfo {
+
+        public SummaryDiagnostic() {
+            addChild("eligble", new SetDiagnosticInfo(eligible));
+            addChild("inprogress", new SetDiagnosticInfo(inProgress));
+            addChild("stale", new SetDiagnosticInfo(stale));
+        }
+
+        public long getEligibleSize() {
+            return eligible.size();
+        }
+
+        public long getInProgressSize() {
+            return inProgress.size();
+        }
+
+        public long getStaleSize() {
+            return stale.size();
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.



**Problem**  
 This line may not be thread-safe: **213**. At this line the following field is not protected by synchronization: `stale`. Your code typically protects this field with synchronization on object: `TrackingSet.this`.

**Fix**  
 Consider protecting the above field with synchronization on object: `TrackingSet.this`. For example, like at lines: **116**, **88**, **95**, **81**, **74**, and **153**.

**More info**  
 The following lines are where the above field is already protected by synchronization on the above object: **56**, **74**, **81**, **88**, **95**, **116**, **123**, and **153**. This is a call to `size()`. You are already protecting `size()` and `isEmpty()`. You are protecting these operations at lines: **88**, **95**, and **116**.